woocommerce/woocommerce
# [COT] Implement `query()` in the COT datastore

### Changes proposed in this Pull Request:
This PR implements the `query()` method in the COT datastore by means of a new `OrdersTableQuery` class (and helper `OrdersTableMetaQuery` for handling meta queries).

- Some properties still need to be added to the querying mechanism as they haven't been fully implemented in the schema yet (`type`, [for example](https://github.com/woocommerce/woocommerce/issues/33771)). When those issues are addressed, we'll have to remember to update `OrdersTableQuery` too.
- Support for legacy query vars, extensibility, as well as some special handling for meta that is now stored in table columns is in the scope of other issues (see #33877, #33878, #33879). This PR deals with the basics of the `query()` implementation, supporting some of the core columns (across the various tables) and also includes a full implementation of meta queries, with the same flexibility offered by WP's own [`meta_query`](https://developer.wordpress.org/reference/classes/wp_meta_query/).
- Some query vars from [`WP_Query`](https://developer.wordpress.org/reference/classes/wp_query/) are supported and automatically translated to COT columns (examples are `post__in` or `post_date`).

Closes #33613 and #33876.

### How to test the changes in this Pull Request:

1. Enable the COT feature and make COT authoritative.
2. Migrate some orders over to COT to make sure we have data to work with.
3. Test the querying functionality via custom code:
   ```php
   $results = wc_get_orders( $args );
   // ... where `$args` is an array of query vars.
   ```

   <details>
      <summary>👀  More examples</summary>

      ```php
      // Orders  with customer_id = 1.
      $results = wc_get_orders(
	      array(
		      'customer_id' => 1,
	      )
      );
      
      // IDs of 10 on-hold orders.
      $results = wc_get_orders(
	      array(
		      'status'   => 'on-hold',
		      'limit'    => 10,
		      'return'   => 'ids',
	      )
      );
      
      // All orders (except for those with ID=1,2,3) with any status, a parent equal to 0 or 1, created via 'checkout' and with meta key my_meta=yes.
      $results = wc_get_orders(
	      array(
		      'status'      => 'all',
		      'exclude'     => array( 1, 2, 3 ),
		      'parent'      => array( 0, 1 ),
		      'meta_key'    => 'my_meta',
		      'meta_value'  => 'yes',
		      'created_via' => 'checkout',
	      )
      );
      
      // A more complex meta_query: orders where 'my_meta'='good' and either 'your_meta'='bad' or 'our_meta' is set (to any value).
      $results = wc_get_orders(
	      array(
		      'meta_query' => array(
			      'relation' => 'AND',
			      array(
				      'key'   => 'my_meta',
				      'value' => 'good',
			      ),
			      array(
				      'relation' => 'OR',
				      array(
					      'key'   => 'your_meta',
					      'value' => 'bad',
				      ),
				      array(
					      'key'   => 'our_meta'
				      )
			      )
		      ),
	      )
      );
      
      // Orders where meta key 'location' is set to something LIKE "London" and meta key 'no_meta' does not exist.
      $results = wc_get_orders(
	      array(
		      'meta_query' => array(
			      array(
				      'key'     => 'location',
				      'value'   => 'London',				
				      'compare' => 'LIKE',
			      ),
			      array(
				      'key'     => 'no_meta',
				      'compare' => 'NOT EXISTS',
			      )
		      ),
	      )
      );
      ```
   </details>
4. Repeat the above with different arguments (see more examples above). Keep in mind that syntax for the `meta_query` argument is the same as for WP's `WP_Query`. See [the docs](https://developer.wordpress.org/reference/classes/wp_query/#custom-field-post-meta-parameters).
